### PR TITLE
Use k8s.io/utils/cpuset package instead of internal one

### LINKED ISSUE #10170: Many cpuset related functions and definitions could be replaced with `k8s.io/utils/cpuset` package
**Is your feature request related to a problem? Please describe**:
Currently to parse cpuset files (with format http://man7.org/linux/man-pages/man7/cpuset.7.html#FORMATS) we use https://github.com/kubevirt/kubevirt/blob/v1.0.0/pkg/util/hardware/hw_utils.go#L42-L73. Exactly the same functionally can be achieved by using https://pkg.go.dev/k8s.io/utils/cpuset#Parse function. This also returns a struct `CPUSet` which we could use to easily manipulate and check various things, right now we pass list of ints (`[]int`).

**Describe the solution you'd like**:
Make use of `k8s.io/utils/cpuset` in places where we manipulate and do things with cpusets/cpulists.

**Additional context**:
https://github.com/kubernetes/utils/pull/267 started to live as separate package in `k8s.io/utils` as it was taken out from `k8s.io/kubernetes/.../kubevirt/.../cpuset`.

### PR DESCRIPTION
**Pull request by VirrageS**
**What this PR does / why we need it**: This PR replaces our own custom function to parse cpulist with `k8s.io/utils/cpuset` package. Thanks to this we can offload implementation and rely on package provided by Kubernetes team.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10170

**Special notes for your reviewer**: None

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-handler/node-labeller/api/capabilities.go [hunk 1]
```diff
@@ -107,8 +107,8 @@ func (b *yesnobool) UnmarshalXMLAttr(attr xml.Attr) error {
 
 func (b *CPUSiblings) UnmarshalXMLAttr(attr xml.Attr) error {
 	if attr.Value != "" {
-		if list, err := hwutil.ParseCPUSetLine(attr.Value, 100); err == nil {
-			for _, cpu := range list {
+		if cpuSet, err := cpuset.Parse(attr.Value); err == nil {
+			for _, cpu := range cpuSet.List() {
```
Comment (fossedihelm) [inline pkg/virt-handler/node-labeller/api/capabilities.go:111]: nit: Is there any particular reason to have the list sorted?

**Review by fossedihelm — COMMENTED** [review]
Thank you! Very little comment below. :)

File: pkg/virt-handler/node-labeller/api/capabilities.go (hunk 1 above)
Comment (VirrageS) [reply pkg/virt-handler/node-labeller/api/capabilities.go:111]: Good question :D This is unmarshaing so probably doesn't really make sense. Let me know if I should change to `UnsortedList()`.

Comment (fossedihelm) [reply pkg/virt-handler/node-labeller/api/capabilities.go:111]: As per offline discussion, it has been decided to always use `List()`.
Thanks

**Comment by rthallisey** [comment]
/retest
@xpivarc can you review

**Comment by xpivarc** [comment]
There are a few points to consider here:
1. Maintenance of the function that we replace is `limit -> 0` (none at this point) and likely will be as Kernel doesn't change interfaces that much.
2. On the other hand the dependency can be updated or we may need to update because of a CVE or anything else.
3. The library does more than we need
4. Our goal is to keep the memory footprint of `virt-launcher` as low as we can. I have a feeling this is going slightly against this goal. Overall Kubernetes packages are bulky. For example, see https://github.com/kubevirt/kubevirt/pull/9079 

Because of the above points, I would be against this PR.

**Comment by jean-edouard** [comment]
> There are a few points to consider here:
> 
> 1. Maintenance of the function that we replace is `limit -> 0` (none at this point) and likely will be as Kernel doesn't change interfaces that much.
> 2. On the other hand the dependency can be updated or we may need to update because of a CVE or anything else.
> 3. The library does more than we need
> 4. Our goal is to keep the memory footprint of `virt-launcher` as low as we can. I have a feeling this is going slightly against this goal. Overall Kubernetes packages are bulky. For example, see [Reduce virt-launcher-monitor binary size #9079](https://github.com/kubevirt/kubevirt/pull/9079)
> 
> Because of the above points, I would be against this PR.

I don't have a strong opinion for this particular case, but I don't like the precedent set by rejecting it.
IMO, using maintained libraries instead of writing and maintaining our own functions is almost always better. Exceptions include, to your 4th point, libs that declare massive global consts (like https://github.com/kubevirt/kubevirt/pull/7414). I don't think it applies here (TBD though).
The potential CVE you mention in 2. is a good point, as parsing functions are particularly susceptible to exploits, but that's an argument in favor of using external packages IMO!

@xpivarc If the author of this PR compared memory footprints with and without this change and found negligible impact, would you reconsider approving it?

**Comment by xpivarc** [comment]
> > There are a few points to consider here:
> > 
> > 1. Maintenance of the function that we replace is `limit -> 0` (none at this point) and likely will be as Kernel doesn't change interfaces that much.
> > 2. On the other hand the dependency can be updated or we may need to update because of a CVE or anything else.
> > 3. The library does more than we need
> > 4. Our goal is to keep the memory footprint of `virt-launcher` as low as we can. I have a feeling this is going slightly against this goal. Overall Kubernetes packages are bulky. For example, see [Reduce virt-launcher-monitor binary size #9079](https://github.com/kubevirt/kubevirt/pull/9079)
> > 
> > Because of the above points, I would be against this PR.
> 
> I don't have a strong opinion for this particular case, but I don't like the precedent set by rejecting it. IMO, using maintained libraries instead of writing and maintaining our own functions is almost always better. Exceptions include, to your 4th point, libs that declare massive global consts (like #7414). I don't think it applies here (TBD though). 

What constant are you referring to? 

The potential CVE you mention in 2. is a good point, as parsing functions are particularly susceptible to exploits, but that's an argument in favor of using external packages IMO!
> 
> @xpivarc If the author of this PR compared memory footprints with and without this change and found negligible impact, would you reconsider approving it?

Yes, in this case, I am happy to reevaluate. However, I am not the sole approver who has a right to merge this into the project. I was just answering a request to review. I am just saying that dependencies have cost as well and a well-written code (& tested!) should not be the main goal of anyone's efforts. We do have many issues to work on (including code that needs re-writing) and I think these issues should be addressed with higher priority.

**Comment by xpivarc** [comment]
> > > > There are a few points to consider here:
> > > > 
> > > > 1. Maintenance of the function that we replace is `limit -> 0` (none at this point) and likely will be as Kernel doesn't change interfaces that much.
> > > > 2. On the other hand the dependency can be updated or we may need to update because of a CVE or anything else.
> > > > 3. The library does more than we need
> > > > 4. Our goal is to keep the memory footprint of `virt-launcher` as low as we can. I have a feeling this is going slightly against this goal. Overall Kubernetes packages are bulky. For example, see [Reduce virt-launcher-monitor binary size #9079](https://github.com/kubevirt/kubevirt/pull/9079)
> > > > 
> > > > Because of the above points, I would be against this PR.
> > > 
> > > 
> > > I don't have a strong opinion for this particular case, but I don't like the precedent set by rejecting it. IMO, using maintained libraries instead of writing and maintaining our own functions is almost always better. Exceptions include, to your 4th point, libs that declare massive global consts (like #7414). I don't think it applies here (TBD though).
> > 
> > 
> > What constant are you referring to?
> 
> https://github.com/subgraph/libmacouflage/blob/master/ouis_bindata.go#L67 (Not a const actually)
>

I remember this one. Yes, that's a good example of when dependency was just too much.

> > > The potential CVE you mention in 2. is a good point, as parsing functions are particularly susceptible to exploits, but that's an argument in favor of using external packages IMO!
> > > @xpivarc If the author of this PR compared memory footprints with and without this change and found negligible impact, would you reconsider approving it?
> > 
> > 
> > Yes, in this case, I am happy to reevaluate. However, I am not the sole approver who has a right to merge this into the project. I was just answering a request to review. I am just saying that dependencies have cost as well and a well-written code (& tested!) should not be the main goal of anyone's efforts. We do have many issues to work on (including code that needs re-writing) and I think these issues should be addressed with higher priority.
> 
> 👍 totally agree about priorities, a bit less on "well tested" (just 2 unit tests for our function vs 8 for the k8s.io one). 

Note: while it is 2 vs 8, 7 out of 8 tests seem to be irrelevant for us. I admit I see some negative test cases we do not have. I just have a hard time imagining myself not approving a PR that would bring this simple function back with full coverage and remove the dependency. Simplicity is great.

>@VirrageS would it be possible to ensure this PR doesn't incur any significant increase in memory usage? Thanks

**Pull request closed without merging**